### PR TITLE
Add TDD: Support for Kafka Connect Source Connectors [HZ-1804]

### PR DESCRIPTION
**Pull request by ldziedziul**
Describe how to add support for Kafka Connect Source Connectors

Fixes https://hazelcast.atlassian.net/browse/HZ-1804

### CODE REVIEW

**Review by orcunc — APPROVED** [review]
This is a general comment. You don't have to do it but I think it would be good for the user to compare when to use CDC/Debezium connectors and Kafka Connect briefly
For example 
1. Kafka Connect may be preferred if DB is not set up for CDC
2. CDC can be source only, but Kafka Connect can be both source and sink
3. ...

**Comment by ldziedziul** [comment]
> This is a general comment. You don't have to do it but I think it would be good for the user to compare when to use CDC/Debezium connectors and Kafka Connect briefly For example
> 1. Kafka Connect may be preferred if DB is not set up for CDC

It depends on the connector and a use case, e.g. Generic JDBC connector periodically executes a SQL query and creates an output record for each row in the result set, so it's rather a batch-like activity. CDC on the other hand sends single changes in the realtime

> 2. CDC can be source only, but Kafka Connect can be both source and sink

For now we support only KC source conncetors

File: docs/design/jet/023-kafka-connect-source-connectors.md [hunk 1]
```diff
@@ -0,0 +1,340 @@
...
+Everything is orchestrated by the `SourceBuilder`
+methods: `createFn`, `fillBufferFn`, `createSnapshotFn`, `restoreSnapshotFn`, `destroyFn`.
+
+The current implementation uses `SourceBuilder` to create a stream source of `SourceRecord`.
+`SourceRecord` is Kafka data type returned by the Kafka Source Connectors for further processing.
+
+The plan is to move the code to the main hazelcast repository and make it production ready with the following steps:
+
+- Move Kafka Connect to Hazelcast Extensions module
+- Add Kafka connect source scaling (currently Hazelcast Platform will instantiate a single task for the specified source
```
Comment (TomaszGaweda) [inline docs/design/jet/023-kafka-connect-source-connectors.md:179]: Personally I'd add snapshotting support as a separate point. Not as easy as I thought ;)

Comment (ldziedziul) [reply docs/design/jet/023-kafka-connect-source-connectors.md:178]: already there: https://github.com/hazelcast/hazelcast/pull/23303/files/b002cf0da1f0357cddf78b4e5b20f3ade8e555d1#diff-4e57279648cfda96eda7b4645a2441607d164df5c44b8fc89f89cff1a44fd844R195-R199

**Comment by ldziedziul** [comment]
run-lab-run

**Review by TomaszGaweda — APPROVED** [review]

File: docs/design/jet/023-kafka-connect-source-connectors.md [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+- Attract new customers by making migration from Kafka easier and by providing more ways of integration
+  with other 3rd party systems.
+- Potentially we could free customers using Kafka Connect from Kafka Infrastructure
+
+#### Terminology
+
+| Term                   | Definition                                                                                                                                                                        |
+|------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+| Kafka Source Connector | Kafka Connect component ingesting data from an external system into Kafka topics.                                                                                                 |
+| Kafka Source Connector | Kafka Connect component writing data from Kafka topics to external systems.                                                                                                       |
```
Comment (nandankidambi) [inline docs/design/jet/023-kafka-connect-source-connectors.md:93]: Shouldnt this be Kafka Sink Connector ?

File: docs/design/jet/023-kafka-connect-source-connectors.md [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+#### Questions about the change
+
+- What components in Hazelcast need to change? How do they change?
+  * N/A
+
+- How does this work in an on-prem deployment?
+  * N/A
+
+- How about on AWS and Kubernetes, platform operator?
+  * As long as necessary jars are on the classpath it doesn't have deployment requirements.
```
Comment (nandankidambi) [inline docs/design/jet/023-kafka-connect-source-connectors.md:209]: For  new HZ instance, dont we need to include any of the kafka connect components in the YAML file ?
For a running / existing HZ member, can you add Kafka connect component - dynamically without having to restart ?
How do you deploy these configurations from one environment to other ?

**Pull request merged**